File: python/deepstrike/runtime/kernel_event_log.py
from __future__ import annotations

from typing import Any, Callable, Literal

KernelEventCategory = Literal["syscall", "sched", "mm", "proc", "ipc"]


def category_for_kind(kind: str) -> KernelEventCategory:
    if kind in ("tool_gated", "capability_changed"):
        return "syscall"
    if kind in (
        "compressed",
        "page_out",
        "page_in",
        "page_in_requested",
        "renewed",
        "context_renewed",
        "large_result_spooled",
        "memory_written",
        "memory_queried",
        "memory_validation_failed",
    ):
        return "mm"
    if kind == "agent_process_changed":
        return "proc"
    if kind == "signal_disposed":
        return "ipc"
    return "sched"


KernelPrimitive = Literal["syscall", "sched", "mm"]


def primitive_for_category(category: KernelEventCategory) -> KernelPrimitive:
    if category == "syscall":
        return "syscall"
    if category == "mm":
        return "mm"
    return "sched"


def primitive_for_kind(kind: str) -> KernelPrimitive:
    return primitive_for_category(category_for_kind(kind))


def with_category(event: dict[str, Any]) -> dict[str, Any]:
    category = category_for_kind(event["kind"])
    return {
        **event,
        "category": category,
        "primitive": primitive_for_category(category),
    }


def kernel_observation_to_session_event(
    obs: dict[str, Any],
    turn: int,
    *,
    next_archive_start: int = 0,
    latest_seq: int | None = None,
    archive_ref: str | None = None,
    preserved_refs: list[str] | None = None,
    compression_action: Callable[[str | None], str | None] | None = None,
    spool_ref: str | None = None,
) -> dict[str, Any] | None:
    t = obs.get("turn") or turn
    to_action = compression_action or (lambda _a: None)

    kind = obs.get("kind")
    if kind == "compressed":
        latest = latest_seq if latest_seq is not None else -1
        if latest < next_archive_start:
            return None
        summary = obs.get("summary")
        return with_category({
            "kind": "compressed",
            "turn": t,
            "archived_seq_range": (next_archive_start, latest),
            "action": to_action(obs.get("action")),
            "summary": summary,
            "summary_tokens": max(1, len(summary) // 4) if summary else None,
            "archive_ref": archive_ref,
            "preserved_refs": preserved_refs or [],
        })
    if kind == "renewed":
        return with_category({
            "kind": "context_renewed",
            "turn": t,
            "sprint": obs.get("sprint") or 0,
            "handoff_ref": "",
        })
    if kind == "rollbacked":
        return with_category({
            "kind": "rollbacked",
            "turn": t,
            "checkpoint_history_len": obs.get("checkpoint_history_len") or 0,
            "reason": obs.get("reason"),
        })
    if kind == "capability_changed":
        ev: dict[str, Any] = with_category({
            "kind": "capability_changed",
            "turn": t,
            "added": obs.get("added") or [],
            "removed": obs.get("removed") or [],
        })
        for key in ("change_kind", "capability_id", "version", "mounted_by", "mount_reason"):
            if obs.get(key) is not None:
                ev[key] = obs[key]
        return ev
    if kind == "milestone_advanced":
        return with_category({
            "kind": "milestone_advanced",
            "turn": t,
            "phase_id": obs.get("phase_id") or "",
            "capabilities_unlocked": obs.get("capabilities_unlocked") or [],
        })
    if kind == "milestone_blocked":
        return with_category({
            "kind": "milestone_blocked",
            "turn": t,
            "phase_id": obs.get("phase_id") or "",
            "reason": obs.get("reason") or "",
        })
    if kind == "checkpoint_taken":
        return with_category({
            "kind": "checkpoint_taken",
            "turn": t,
            "history_len": obs.get("history_len") or 0,
        })
    if kind == "agent_process_changed":
        ev = with_category({
            "kind": "agent_process_changed",
            "turn": t,
            "agent_id": obs.get("agent_id") or "",
            "parent_session_id": obs.get("parent_session_id") or "",
            "role": obs.get("role") or "",
            "isolation": obs.get("isolation") or "",
            "context_inheritance": obs.get("context_inheritance") or "",
            "state": obs.get("state") or "running",
            "permitted_capability_ids": obs.get("permitted_capability_ids") or [],
        })
        if obs.get("result_termination"):
            ev["result_termination"] = obs["result_termination"]
        return ev
    if kind == "tool_gated":
        return with_category({
            "kind": "tool_gated",
            "turn": t,
            "call_id": obs.get("call_id") or "",
            "tool": obs.get("tool") or "",
            "reason": obs.get("reason") or "",
        })
    if kind == "signal_disposed":
        return with_category({
            "kind": "signal_disposed",
            "turn": t,
            "signal_id": obs.get("signal_id") or "",
            "disposition": obs.get("disposition") or "",
            "queue_depth": obs.get("queue_depth") or 0,
        })
    if kind == "budget_exceeded":
        return with_category({
            "kind": "budget_exceeded",
            "turn": t,
            "budget": obs.get("budget") or "",
        })
    if kind == "suspended":
        return with_category({
            "kind": "suspended",
            "turn": t,
            "reason": obs.get("reason") or "",
            "pending_calls": obs.get("pending_calls") or [],
        })
    if kind == "resumed":
        return with_category({
            "kind": "resumed",
            "turn": t,
            "approved": obs.get("approved") or [],
            "denied": obs.get("denied") or [],
        })
    if kind == "page_in_requested":
        return None
    if kind == "large_result_spooled":
        return with_category({
            "kind": "large_result_spooled",
            "turn": t,
            "call_id": obs.get("call_id") or "",
            "tool": obs.get("tool") or "",
            "original_size": obs.get("original_size") or 0,
            "preview_size": obs.get("preview_size") or 0,
            "spool_ref": spool_ref,
        })
    if kind == "memory_written":
        return with_category({
            "kind": "memory_written",
            "turn": t,
            "memory_id": obs.get("memory_id") or "",
            "memory_kind": obs.get("memory_kind") or "",
            "size_bytes": obs.get("size_bytes") or 0,
        })
    if kind == "memory_queried":
        return with_category({
            "kind": "memory_queried",
            "turn": t,
            "query_context": obs.get("query_context") or "",
            "requested_k": obs.get("requested_k") or 0,
            "requires_async_response": obs.get("requires_async_response") or False,
        })
    if kind == "memory_validation_failed":
        return with_category({
            "kind": "memory_validation_failed",
            "turn": t,
            "memory_id": obs.get("memory_id") or "",
            "error": obs.get("error") or "",
        })
    return None

File: node/src/runtime/kernel-event-log.ts
import type { KernelObservation } from "./kernel-step.js"
import type { RollbackReason, SessionEvent } from "./session-log.js"

/** Agent OS kernel event category (Phase 5). */
export type KernelEventCategory = "syscall" | "sched" | "mm" | "proc" | "ipc"

export function categoryForKind(kind: string): KernelEventCategory {
  switch (kind) {
    case "tool_gated":
    case "capability_changed":
      return "syscall"
    case "compressed":
    case "page_out":
    case "page_in":
    case "page_in_requested":
    case "renewed":
    case "context_renewed":
    case "large_result_spooled":
    case "memory_written":
    case "memory_queried":
    case "memory_validation_failed":
      return "mm"
    case "agent_process_changed":
      return "proc"
    case "signal_delivery_disposed":
      return "ipc"
    default:
      return "sched"
  }
}

type CompressionAction = Extract<SessionEvent, { kind: "compressed" }>["action"]

export function kernelObservationToSessionEvent(
  obs: KernelObservation,
  turn: number,
  opts: {
    nextArchiveStart?: number
    latestSeq?: number
    preservedRefs?: string[]
    compressionAction?: (action?: string) => CompressionAction
  } = {},
): SessionEvent | null {
  const t = obs.turn ?? turn
  const compressionAction = opts.compressionAction ?? (() => undefined)

  switch (obs.kind) {
    case "compressed": {
      const latest = opts.latestSeq ?? -1
      const start = opts.nextArchiveStart ?? 0
      if (latest < start) return null
      return {
        kind: "compressed" as const,
        turn: t,
        archived_seq_range: [start, latest] as [number, number],
        action: compressionAction(obs.action),
        summary: obs.summary,
        summary_tokens: obs.summary ? Math.max(1, Math.ceil(obs.summary.length / 4)) : undefined,
        preserved_refs: opts.preservedRefs ?? [],
      }
    }
    case "renewed":
      return {
        kind: "context_renewed" as const,
        turn: t,
        sprint: obs.sprint ?? 0,
        handoff_ref: "",
      }
    case "rollbacked":
      return {
        kind: "rollbacked" as const,
        turn: t,
        checkpoint_history_len: obs.checkpoint_history_len ?? 0,
        reason: obs.reason as RollbackReason | undefined,
      }
    case "capability_changed":
      return {
        kind: "capability_changed" as const,
        turn: t,
        added: obs.added ?? [],
        removed: obs.removed ?? [],
        ...(obs.change_kind != null && { change_kind: obs.change_kind }),
        ...(obs.capability_id != null && { capability_id: obs.capability_id }),
        ...(obs.version != null && { version: obs.version }),
        ...(obs.mounted_by != null && { mounted_by: obs.mounted_by }),
        ...(obs.mount_reason != null && { mount_reason: obs.mount_reason }),
      }
    case "milestone_advanced":
      return {
        kind: "milestone_advanced" as const,
        turn: t,
        phase_id: obs.phase_id ?? "",
        capabilities_unlocked: obs.capabilities_unlocked ?? [],
      }
    case "milestone_blocked":
      return {
        kind: "milestone_blocked" as const,
        turn: t,
        phase_id: obs.phase_id ?? "",
        reason: typeof obs.reason === "string" ? obs.reason : "",
      }
    case "checkpoint_taken":
      return {
        kind: "checkpoint_taken" as const,
        turn: t,
        history_len: obs.history_len ?? 0,
      }
    case "entropy_sample":
      return {
        kind: "entropy_sample" as const,
        turn: t,
        score: obs.score ?? 0,
        score_version: obs.score_version ?? 0,
        rho: obs.rho ?? 0,
        repeat_pressure: obs.repeat_pressure ?? 0,
        failure_rate: obs.failure_rate ?? 0,
        rollbacks_in_window: obs.rollbacks_in_window ?? 0,
        window_turns: obs.window_turns ?? 0,
      }
    case "entropy_alert":
      return {
        kind: "entropy_alert" as const,
        turn: t,
        score: obs.score ?? 0,
        threshold: obs.threshold ?? 0,
      }
    case "agent_process_changed":
      return {
        kind: "agent_process_changed" as const,
        turn: t,
        agent_id: obs.agent_id ?? "",
        parent_session_id: obs.parent_session_id ?? "",
        role: obs.role ?? "",
        isolation: obs.isolation ?? "",
        context_inheritance: obs.context_inheritance ?? "",
        state: (obs as { state?: string }).state ?? "running",
        permitted_capability_ids: obs.permitted_capability_ids ?? [],
        ...((obs as { result_termination?: string }).result_termination
          ? { result_termination: (obs as { result_termination?: string }).result_termination }
          : {}),
      }
    case "tool_gated":
      return {
        kind: "tool_gated" as const,
        turn: t,
        call_id: obs.call_id ?? "",
        tool: obs.tool ?? "",
        reason: typeof obs.reason === "string" ? obs.reason : "",
      }
    case "signal_delivery_disposed":
      return {
        kind: "signal_delivery_disposed" as const,
        turn: t,
        operation_id: obs.operation_id ?? "",
        delivery_id: obs.delivery_id ?? "",
        attempt: obs.attempt ?? 0,
        signal_id: obs.signal_id ?? "",
        disposition: obs.disposition ?? "",
        queue_depth: obs.queue_depth ?? 0,
      }
    case "budget_exceeded":
      return {
        kind: "budget_exceeded" as const,
        turn: t,
        budget: obs.budget ?? "",
        operation_id: obs.operation_id ?? "",
        ...(obs.reservation_id ? { reservation_id: obs.reservation_id } : {}),
      }
    case "budget_usage_reported":
      return {
        kind: "budget_usage_reported" as const,
        turn: t,
        operation_id: obs.operation_id ?? "",
        reservation_id: obs.reservation_id ?? "",
        tokens: obs.tokens ?? 0,
        subagents: obs.subagents ?? 0,
        rounds: obs.rounds ?? 0,
      }
    case "operation_cancelled":
      return {
        kind: "operation_cancelled" as const,
        turn: t,
        operation_id: obs.operation_id ?? "",
        reason: (obs.reason ?? "user") as "user" | "deadline" | "lease_lost" | "host_shutdown",
        pending_call_ids: obs.pending_call_ids ?? [],
      }
    case "suspended":
      return {
        kind: "suspended" as const,
        turn: t,
        reason: typeof obs.reason === "string" ? obs.reason : "",
        pending_calls: obs.pending_calls ?? [],
      }
    case "resumed":
      return {
        kind: "resumed" as const,
        turn: t,
        approved: obs.approved ?? [],
        denied: obs.denied ?? [],
      }
    case "page_in_requested":
      return null
    case "large_result_spooled":
      return {
        kind: "large_result_spooled" as const,
        turn: t,
        call_id: obs.call_id ?? "",
        tool: obs.tool ?? "",
        original_size: obs.original_size ?? 0,
        preview_size: obs.preview_size ?? 0,
        spool_ref: obs.spool_ref,
      }
    case "page_out_archived":
      return {
        kind: "page_out" as const,
        turn: t,
        action: compressionAction(obs.action),
        summary: obs.summary,
        tier_hint: obs.tier,
        message_count: obs.message_count ?? 0,
        archive_ref: obs.archive_ref,
      }
    case "memory_written":
      return {
        kind: "memory_written" as const,
        turn: t,
        record_id: obs.record_id ?? "",
        scope: obs.scope ?? { tenant_id: "", namespace: "" },
        memory_kind: obs.memory_kind ?? "",
        name: obs.name ?? "",
        size_bytes: obs.size_bytes ?? 0,
      }
    case "memory_queried":
      return {
        kind: "memory_queried" as const,
        turn: t,
        scope: obs.scope ?? { tenant_id: "", namespace: "" },
        query: obs.query ?? "",
        requested_k: obs.requested_k ?? 0,
        requires_async_response: obs.requires_async_response ?? false,
      }
    case "memory_validation_failed":
      return {
        kind: "memory_validation_failed" as const,
        turn: t,
        record_id: obs.record_id ?? "",
        error: obs.error ?? "",
      }
    case "memory_write_failed":
      return {
        kind: "memory_write_failed" as const,
        turn: t,
        record_id: obs.record_id ?? "",
        error: obs.error ?? "",
      }
    case "memory_query_failed":
      return {
        kind: "memory_query_failed" as const,
        turn: t,
        scope: obs.scope ?? { tenant_id: "", namespace: "" },
        query: obs.query ?? "",
        error: obs.error ?? "",
      }
    case "workflow_batch_spawned": {
      // Batch metadata persisted for resume recovery; individual nodes are
      // recorded when they complete (via workflow_node_completed).
      const nodes = (obs as any).nodes ?? []
      return {
        kind: "workflow_batch_spawned" as const,
        turn: t,
        node_count: nodes.length,
        node_ids: nodes.map((n: any) => n.agent_id ?? ""),
      }
    }
    case "workflow_completed": {
      const nodeOutcomes = (obs as any).node_outcomes ?? []
      return {
        kind: "workflow_completed" as const,
        turn: t,
        node_outcomes: nodeOutcomes,
        total_nodes: nodeOutcomes.length,
      }
    }
    default:
      return null
  }
}

export type KernelPrimitive = "syscall" | "sched" | "mm"

export function primitiveForCategory(category: KernelEventCategory): KernelPrimitive {
  switch (category) {
    case "syscall":
      return "syscall"
    case "mm":
      return "mm"
    case "proc":
    case "ipc":
    case "sched":
    default:
      return "sched"
  }
}

export function primitiveForKind(kind: string): KernelPrimitive {
  return primitiveForCategory(categoryForKind(kind))
}
